pkg/state/infra.go

```diff
@@ -116,24 +195,319 @@ func (i *infraClient) SyncConnectors(routers []string, brokers []string) ([]Conn
+	for _, internalPort := range endpoint.Status.InternalPorts {
+		if internalPort.Protocol == v1beta2.MessagingProtocolAMQP {
+			err := i.applyRouters(context.Background(), func(router *RouterState) error {
+				return router.EnsureListener(&RouterListener{
+					Name:               internalPort.Name,
+					Host:               "0.0.0.0",
+					Port:               fmt.Sprintf("%d", internalPort.Port),
+					Role:               "normal",
+					IdleTimeoutSeconds: 16,
+					MultiTenant:        true,
+				})
+			})
+			if err != nil {
+				return err
+			}
+		} else {
+			// TODO
+			return fmt.Errorf("%s protocol not yet supported", internalPort.Protocol)
+		}
+	}
+	return nil
+}
+
+func (i *infraClient) SyncAddress(address *v1beta2.MessagingAddress) error {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	synced := 0
+	for _, endpoint := range i.endpoints {
+		if endpoint.Status.Phase == v1beta2.MessagingEndpointActive && endpoint.Status.Host != "" {
+			err := i.syncAddressInternal(endpoint, address)
+			if err != nil {
+				return err
+			}
+			synced++
+		}
+	}
+	if synced == 0 {
+		return fmt.Errorf("No active endpoints defined")
+	}
+	i.addresses[resourceKey{Name: address.Name, Namespace: address.Namespace}] = address
+	return nil
+}
+
+func (i *infraClient) syncAddressInternal(endpoint *v1beta2.MessagingEndpoint, address *v1beta2.MessagingAddress) error {
+	ctx := context.Background()
+
+	// Skip endpoints that are not active or do not have hosts defined
+	if endpoint.Status.Phase != v1beta2.MessagingEndpointActive || endpoint.Status.Host == "" {
+		return nil
+	}
+
+	tenantId := endpoint.Status.Host
+
+	// Configure all routers
+	err := i.applyRouters(ctx, func(router *RouterState) error {
+		addressName := addressName(tenantId, address)
+		if address.Spec.Anycast != nil {
+			err := router.EnsureAddress(&RouterAddress{
+				Name:         addressName,
+				Prefix:       fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+				Distribution: "balanced",
+				Waypoint:     false,
+			})
+			if err != nil {
+				return err
+			}
+		} else if address.Spec.Multicast != nil {
+			err := router.EnsureAddress(&RouterAddress{
+				Name:         addressName,
+				Prefix:       fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+				Distribution: "multicast",
+				Waypoint:     false,
+			})
+			if err != nil {
+				return err
+			}
+		} else if address.Spec.Queue != nil {
+			err := router.EnsureAddress(&RouterAddress{
+				Name:         addressName,
+				Prefix:       fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+				Distribution: "balanced",
+				Waypoint:     true,
+			})
+			if err != nil {
+				return err
+			}
+
+			for _, broker := range address.Status.Brokers {
+				brokerState := i.brokers[broker.Host]
+				if brokerState == nil {
+					return fmt.Errorf("Unable to configure address autoLink (tenant %s address %s) for unknown broker %s", tenantId, address.GetAddress(), broker.Host)
```

Comment:
The Go advice on capitalisation at the start of error messages.  https://github.com/golang/go/wiki/CodeReviewComments#error-strings